refactor(ast_codegen): simplify `derive_get_span` generator

overlookmotel opened the pull request:
Simplify `derive_get_span` generator that was introduced in #4735. No change to functionality, just aiming for greater readability.

In particular:

* Move defining idents/tokens which are specific to `GetSpan` / `GetSpanMut` into those specific generators, rather than branching on `MUT` later on.
* Remove `const MUT` param.
* Remove the confusing pairs of closures and functions both called `derive_enum` / `derive_struct`.
* Inline function which generates the impls - prioritizing readability over DRY code.

overlookmotel commented:
What do you think @rzvxa? Personally I find this style less convoluted and a lot easier to understand.

NB: Rather than looking at the diff, I'd suggest comparing before and after.

rzvxa commented:
> What do you think @rzvxa? Personally I find this style less convoluted and a lot easier to understand.

TBH, I don't have a strong preference here. I'll merge it in if you like it more this way.

overlookmotel committed:
refactor(ast_codegen): simplify `derive_get_span` generator (#4757)

Simplify `derive_get_span` generator that was introduced in #4735. No change to functionality, just aiming for greater readability.

In particular:

* Move defining idents/tokens which are specific to `GetSpan` / `GetSpanMut` into those specific generators, rather than branching on `MUT` later on.
* Remove `const MUT` param.
* Remove the confusing pairs of closures and functions both called `derive_enum` / `derive_struct`.
* Inline function which generates the impls - prioritizing readability over DRY code.
